hw/hw_06_task_02.py

```diff
@@ -0,0 +1,182 @@
+3. Student и Teacher имеют одинаковые по смыслу атрибуты
+(last_name, first_name) - избавиться от дублирования с помощью наследования
+4.
+Teacher
+Атрибут:
+    homework_done - структура с интерфейсом как в словаря, сюда поподают все
+    HomeworkResult после успешного прохождения check_homework
+    (нужно гаранитровать остутствие повторяющихся результатов по каждому
+    заданию), группировать по экземплярам Homework.
+    Общий для всех учителей. Вариант ипользования смотри в блоке if __main__...
+Методы:
+    check_homework - принимает экземпляр HomeworkResult и возвращает True если
+    ответ студента больше 5 символов, так же при успешной проверке добавить в
+    homework_done.
+    Если меньше 5 символов - никуда не добавлять и вернуть False.
+    reset_results - если передать экземпряр Homework - удаляет только
+    результаты этого задания из homework_done, если ничего не передавать,
+    то полностью обнулит homework_done.
+PEP8 соблюдать строго.
+Всем перечисленным выше атрибутам и методам классов сохранить названия.
+К названием остальных переменных, классов и тд. подходить ответственно -
+давать логичные подходящие имена.
+"""
+
+from collections import defaultdict
+from datetime import datetime, timedelta
+from typing import Optional
+
+
+class DeadlineError(Exception):
+    def __init__(self):
+        super().__init__("You are late.")
+
+
+class Person:
+    """
+    A class to represent a person.
+
+    Attributes
+    ----------
+    last_name : str
+        last name of a person
+    first_name: str
+        first name of a person
+
+    """
+
+    def __init__(self, first_name: str, last_name: str):
+        """Create an instance of a person."""
+        self.last_name = last_name
+        self.first_name = first_name
+
+
+class Homework:
+    """
+    A class to represent a homework.
+
+    Attributes
+    ----------
+    text : str
+        text of homework
+    deadline: int
+        num of days to do homework
+    created: datetime.timedelta
+        date and time of deadline
+
+    """
+
+    def __init__(self, text: str, deadline: int):
+        """Create an instance of a homework."""
+        self.text = text
+        self.deadline = timedelta(deadline)
+        self.created = datetime.now()
+
+    def is_active(self) -> bool:
+        """Return False if deadline is expired, True otherwise."""
+        hw_deadline = self.created + self.deadline
+        now = datetime.now()
+        return not now > hw_deadline
+
+
+class Student(Person):
+    """
+    A class to represent a student.
+
+    Methods
+    -------
+    do_homework():
+        Return homework if deadline is available, raise exception otherwise.
+
+    """
+
+    # flake8: noqa: T001
+    def do_homework(self, hw: Homework, solution: str) -> Optional[Homework]:
+        """Return None and print message if deadline is passed, an instance of homework otherwise."""
+        if not hw.is_active():
```

Comment:
This line has 3 actions:  condition, negation, and function call.  You can reorganize the code and get rid of the negation.

```
if hw.is_active():
    return HomeworkResult(self, hw, solution)
raise DeadlineError()